File: AOC-2015/Day8/day8.cpp
#include <iostream>
#include <fstream>
#include <string>

int calculateStringDifference(const std::string& line) {
    int codeLength = line.length();
    int memoryLength = 0;

    for (size_t i = 1; i < line.length() - 1; ++i) {
        char ch = line[i];
        if (ch == '\\') {
            if (line[i + 1] == '\\' || line[i + 1] == '\"') {
                memoryLength++;
                i++;
            } else if (line[i + 1] == 'x') {
                memoryLength++;
                i += 3;
            } else {
                memoryLength++;
                i++;
            }
        } else {
            memoryLength++;
        }
    }

    return codeLength - memoryLength;
}

std::string encodeString(const std::string& line) {
    std::string encodedLine = "\"";

    for (char ch : line) {
        if (ch == '\\' || ch == '\"') {
            encodedLine += '\\';
        }
        encodedLine += ch;
    }

    encodedLine += '\"';
    return encodedLine;
}

int calculateEncodedDifference(const std::string& line) {
    std::string encodedLine = encodeString(line);
    return encodedLine.length() - line.length();
}

int main() {
    int totalDifference = 0;
    std::ifstream file("input.txt");
    std::string line;
    
    if (file.is_open()) {
        while (std::getline(file, line)) {
            totalDifference += calculateStringDifference(line);
        }
        file.close();
    }

    std::cout << "Total difference: " << totalDifference << std::endl;

    int totalEncodedDifference = 0;
    file.open("input.txt");

    if (file.is_open()) {
        while (std::getline(file, line)) {
            totalEncodedDifference += calculateEncodedDifference(line);
        }
        file.close();
    }

    std::cout << "Total encoded difference: " << totalEncodedDifference << std::endl;

    return 0;
}

File: AOC-2015/Day8/day8.py
def calculate_string_difference(line):
    line = line.strip()

    code_length = len(line)

    memory_length = 0
    index = 1 
    while index < len(line) - 1:
        char = line[index]
        if char == "\\":
            # Handle escape sequences
            if line[index + 1] == "\\" or line[index + 1] == "\"":
                # \\ or \"
                memory_length += 1
                index += 2
            elif line[index + 1] == "x":
                # \x followed by two hexadecimal characters
                memory_length += 1
                index += 4
            else:
                # Single backslash or other escape sequence
                memory_length += 1
                index += 2
        else: # Regular character
            memory_length += 1
            index += 1

    return code_length - memory_length

def encode_string(line):
    encoded_line = "\""

    for char in line:
        if char == "\\" or char == "\"":
            # Escape backslash and double quote
            encoded_line += "\\" + char
        elif char == "\n":
            # Skip newline characters
            continue
        else: # Regular character
            encoded_line += char

    # Add closing double quote
    encoded_line += "\""

    return encoded_line

def calculate_encoded_difference(line):
    line = line.strip()
    encoded_length = len(encode_string(line))

    return encoded_length - len(line)

def main():
    total_difference = 0
    with open("input.txt", "r") as file:
        for line in file:
            total_difference += calculate_string_difference(line)

    print("Total difference:", total_difference)

    total_encoded_difference = 0
    with open("input.txt", "r") as file:
        for line in file:
            total_encoded_difference += calculate_encoded_difference(line)

    print("Total encoded difference:", total_encoded_difference)

if __name__ == "__main__":
    main()
